polar-core/src/vm.rs

```diff
@@ -1096,48 +1070,43 @@ impl PolarVirtualMachine {
+
+                // Produce a constraint like left.field = value
+                let to_unify = |(field, value): (&Symbol, &Term)| -> Term {
+                    let value = self.deref(value);
                     let field = right.clone_with_value(value!(field.0.as_ref()));
                     let left = left.clone_with_value(value!(op!(Dot, left.clone(), field)));
-                    let unify = op!(Unify, left.clone(), value.clone());
-                    match value.value() {
-                        Value::Variable(v) | Value::RestVariable(v) => {
-                            match self.variable_state(v) {
-                                VariableState::Cycle(c) => cycle_constraints(c)
-                                    .clone_with_new_constraint(unify.into_term()),
-                                VariableState::Bound(b) => op!(Unify, left, b),
-                                VariableState::Partial(e) => {
-                                    e.clone_with_new_constraint(unify.into_term())
-                                }
-                                VariableState::Unbound => unify,
-                            }
-                        }
-                        _ => unify,
-                    }
+                    let unify = op!(Unify, left, value);
+                    term!(unify)
                 };
 
-                let mut operation = operation.clone();
-
                 // Add all but the last field constraint to the operation.
-                for op in fields.fields.iter().skip(1).rev().map(to_unify) {
-                    operation.add_constraint(op);
+                let constraints = fields.fields.iter().rev().map(to_unify).collect::<Vec<_>>();
+                for op in constraints {
+                    self.add_constraint(&op)?;
                 }
-
-                // Add the last field constraint and trigger the bind dance.
-                let op = fields.fields.iter().take(1).map(to_unify).next().unwrap();
-                self.constrain(&operation.clone_with_new_constraint(op.into_term()))?;
             }
             Value::Pattern(Pattern::Instance(InstanceLiteral { fields, tag })) => {
                 // TODO(gj): assert that a simplified expression contains at most 1 unification
                 // involving a particular variable.
                 // TODO(gj): Ensure `op!(And) matches X{}` doesn't die after these changes.
-
                 let var = left.value().as_symbol()?;
-                let simplified = simplify_partial(var, operation.clone().into_term());
+                // Get the existing partial on the LHS variable.
+                let partial = match self.variable_state(var) {
+                    VariableState::Partial(expr) => expr,
+                    VariableState::Cycle(c) => cycle_constraints(c),
+                    _ => panic!("Invariant violated. left must be a variable"),
+                };
+
+                let simplified = simplify_partial(var, partial.into_term());
```

Comment:
I'm not sure if this is tested or correct.

Reply:
There are some integration tests, but it's not well tested.